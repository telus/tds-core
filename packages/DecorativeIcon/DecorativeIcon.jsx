import React from 'react'
import PropTypes from 'prop-types'

import Icon from '../../shared/components/Icon/Icon'

/**
 * An icon used for visual aesthetics only.
 */
const DecorativeIcon = ({ symbol, variant, size, ...rest }) => (
  <Icon {...rest} symbol={symbol} variant={variant} size={size} aria-hidden="true" />
)

DecorativeIcon.propTypes = {
  /**
   * Name of the icon symbol.
   */
  symbol: PropTypes.oneOf([
    'caretDown',
    'caretUp',
    'checkmark',
    'chevron',
    'leftChevron',
    'exclamationPointCircle',
    'expander',
    'hamburger',
    'location',
    'minus',
    'plus',
    'questionMarkCircle',
    'spyglass',
    'times',
  ]).isRequired,
  /**
   * The appearance of the Icon.
   */
  variant: PropTypes.oneOf(['primary', 'secondary', 'inverted', 'error']),
  /**
   * The icon size in pixels.
   */
  size: PropTypes.oneOf([16, 24, 48]),
}

DecorativeIcon.defaultProps = {
  variant: undefined,
  size: 24,
}

export default DecorativeIcon
