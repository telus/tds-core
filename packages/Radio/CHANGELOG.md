# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [2.0.29](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.28...@tds/core-radio@2.0.29) (2020-07-22)

**Note:** Version bump only for package @tds/core-radio





## [2.0.28](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.27...@tds/core-radio@2.0.28) (2020-04-14)

**Note:** Version bump only for package @tds/core-radio





## [2.0.27](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.26...@tds/core-radio@2.0.27) (2020-03-11)

**Note:** Version bump only for package @tds/core-radio





## [2.0.26](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.25...@tds/core-radio@2.0.26) (2020-03-10)

**Note:** Version bump only for package @tds/core-radio





## [2.0.25](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.24...@tds/core-radio@2.0.25) (2020-03-09)

**Note:** Version bump only for package @tds/core-radio





## [2.0.24](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.23...@tds/core-radio@2.0.24) (2020-02-24)

**Note:** Version bump only for package @tds/core-radio





## [2.0.23](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.22...@tds/core-radio@2.0.23) (2020-02-03)

**Note:** Version bump only for package @tds/core-radio





## [2.0.22](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.21...@tds/core-radio@2.0.22) (2020-01-27)

**Note:** Version bump only for package @tds/core-radio





## [2.0.21](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.20...@tds/core-radio@2.0.21) (2020-01-23)

**Note:** Version bump only for package @tds/core-radio





## [2.0.20](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.19...@tds/core-radio@2.0.20) (2020-01-22)

**Note:** Version bump only for package @tds/core-radio





## [2.0.19](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.18...@tds/core-radio@2.0.19) (2019-11-18)

**Note:** Version bump only for package @tds/core-radio





## [2.0.18](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.17...@tds/core-radio@2.0.18) (2019-10-10)


### Bug Fixes

* **core-radio:** fix proportions with custom browser font sizes ([f78bbc3](https://github.com/telusdigital/tds/commit/f78bbc3))





## [2.0.17](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.16...@tds/core-radio@2.0.17) (2019-10-09)

**Note:** Version bump only for package @tds/core-radio





## [2.0.16](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.15...@tds/core-radio@2.0.16) (2019-10-04)

**Note:** Version bump only for package @tds/core-radio





## [2.0.15](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.14...@tds/core-radio@2.0.15) (2019-09-12)

**Note:** Version bump only for package @tds/core-radio





## [2.0.14](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.13...@tds/core-radio@2.0.14) (2019-09-06)

**Note:** Version bump only for package @tds/core-radio





## [2.0.13](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.12...@tds/core-radio@2.0.13) (2019-08-22)

**Note:** Version bump only for package @tds/core-radio





## [2.0.12](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.11...@tds/core-radio@2.0.12) (2019-08-09)

**Note:** Version bump only for package @tds/core-radio





## [2.0.11](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.10...@tds/core-radio@2.0.11) (2019-08-01)

**Note:** Version bump only for package @tds/core-radio





## [2.0.10](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.9...@tds/core-radio@2.0.10) (2019-07-30)

**Note:** Version bump only for package @tds/core-radio





## [2.0.9](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.8...@tds/core-radio@2.0.9) (2019-07-25)

**Note:** Version bump only for package @tds/core-radio





## [2.0.8](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.7...@tds/core-radio@2.0.8) (2019-07-17)

**Note:** Version bump only for package @tds/core-radio





## [2.0.7](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.6...@tds/core-radio@2.0.7) (2019-07-16)

**Note:** Version bump only for package @tds/core-radio





## [2.0.6](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.5...@tds/core-radio@2.0.6) (2019-07-15)


### Bug Fixes

* **core-radio:** wrap long labels in ie ([7c450b0](https://github.com/telusdigital/tds/commit/7c450b0))





## [2.0.5](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.4...@tds/core-radio@2.0.5) (2019-07-10)

**Note:** Version bump only for package @tds/core-radio





## [2.0.4](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.3...@tds/core-radio@2.0.4) (2019-07-02)

**Note:** Version bump only for package @tds/core-radio





## [2.0.3](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.2...@tds/core-radio@2.0.3) (2019-06-25)

**Note:** Version bump only for package @tds/core-radio





## [2.0.2](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.1...@tds/core-radio@2.0.2) (2019-06-18)

**Note:** Version bump only for package @tds/core-radio





## [2.0.1](https://github.com/telusdigital/tds/compare/@tds/core-radio@2.0.0...@tds/core-radio@2.0.1) (2019-06-04)

**Note:** Version bump only for package @tds/core-radio

# [2.0.0](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.20...@tds/core-radio@2.0.0) (2019-05-23)

### Features

- **core-radio:** upgrade to styled-components ([2e677b0](https://github.com/telusdigital/tds/commit/2e677b0))

### BREAKING CHANGES

- **core-radio:** styled-components now a peer dependency

## [1.2.20](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.19...@tds/core-radio@1.2.20) (2019-05-22)

**Note:** Version bump only for package @tds/core-radio

## [1.2.19](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.18...@tds/core-radio@1.2.19) (2019-05-01)

**Note:** Version bump only for package @tds/core-radio

## [1.2.18](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.17...@tds/core-radio@1.2.18) (2019-04-29)

**Note:** Version bump only for package @tds/core-radio

## [1.2.17](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.16...@tds/core-radio@1.2.17) (2019-04-23)

**Note:** Version bump only for package @tds/core-radio

## [1.2.16](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.15...@tds/core-radio@1.2.16) (2019-04-22)

**Note:** Version bump only for package @tds/core-radio

## [1.2.15](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.14...@tds/core-radio@1.2.15) (2019-04-16)

**Note:** Version bump only for package @tds/core-radio

## [1.2.14](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.13...@tds/core-radio@1.2.14) (2019-04-11)

**Note:** Version bump only for package @tds/core-radio

## [1.2.13](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.12...@tds/core-radio@1.2.13) (2019-04-11)

**Note:** Version bump only for package @tds/core-radio

## [1.2.12](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.11...@tds/core-radio@1.2.12) (2019-03-28)

**Note:** Version bump only for package @tds/core-radio

## [1.2.11](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.10...@tds/core-radio@1.2.11) (2019-03-28)

**Note:** Version bump only for package @tds/core-radio

## [1.2.10](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.9...@tds/core-radio@1.2.10) (2019-03-20)

**Note:** Version bump only for package @tds/core-radio

## [1.2.9](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.8...@tds/core-radio@1.2.9) (2019-03-20)

**Note:** Version bump only for package @tds/core-radio

## [1.2.8](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.7...@tds/core-radio@1.2.8) (2019-03-18)

**Note:** Version bump only for package @tds/core-radio

## [1.2.7](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.6...@tds/core-radio@1.2.7) (2019-03-18)

**Note:** Version bump only for package @tds/core-radio

## [1.2.6](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.5...@tds/core-radio@1.2.6) (2019-03-15)

**Note:** Version bump only for package @tds/core-radio

## [1.2.5](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.4...@tds/core-radio@1.2.5) (2019-03-11)

**Note:** Version bump only for package @tds/core-radio

## [1.2.4](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.3...@tds/core-radio@1.2.4) (2019-03-05)

**Note:** Version bump only for package @tds/core-radio

## [1.2.3](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.2...@tds/core-radio@1.2.3) (2019-02-20)

**Note:** Version bump only for package @tds/core-radio

## [1.2.2](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.1...@tds/core-radio@1.2.2) (2019-02-06)

**Note:** Version bump only for package @tds/core-radio

<a name="1.2.1"></a>

## [1.2.1](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.2.0...@tds/core-radio@1.2.1) (2018-12-19)

**Note:** Version bump only for package @tds/core-radio

<a name="1.2.0"></a>

# [1.2.0](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.1.1...@tds/core-radio@1.2.0) (2018-12-13)

### Features

- **core-radio:** add description text functionality ([dc418c6](https://github.com/telusdigital/tds/commit/dc418c6))

<a name="1.1.1"></a>

## [1.1.1](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.1.0...@tds/core-radio@1.1.1) (2018-12-06)

**Note:** Version bump only for package @tds/core-radio

<a name="1.1.0"></a>

# [1.1.0](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.0.5...@tds/core-radio@1.1.0) (2018-07-23)

### Features

- **core-radio:** allow boolean proptype for value ([199292d](https://github.com/telusdigital/tds/commit/199292d))

<a name="1.0.5"></a>

## [1.0.5](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.0.4...@tds/core-radio@1.0.5) (2018-07-10)

**Note:** Version bump only for package @tds/core-radio

<a name="1.0.4"></a>

## [1.0.4](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.0.3...@tds/core-radio@1.0.4) (2018-06-14)

**Note:** Version bump only for package @tds/core-radio

<a name="1.0.3"></a>

## [1.0.3](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.0.2...@tds/core-radio@1.0.3) (2018-06-07)

### Bug Fixes

- **shared-choice:** call onchange after async setstate, update examples ([4e992b2](https://github.com/telusdigital/tds/commit/4e992b2))

<a name="1.0.2"></a>

## [1.0.2](https://github.com/telusdigital/tds/compare/@tds/core-radio@1.0.1...@tds/core-radio@1.0.2) (2018-04-02)

### Bug Fixes

- **core-box:** update tests ([c0f0343](https://github.com/telusdigital/tds/commit/c0f0343))
