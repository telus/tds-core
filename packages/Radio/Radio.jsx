import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import Box from '@tds/core-box'
import {
  colorCardinal,
  colorWhite,
  colorGreyShuttle,
  colorGreyGainsboro,
  colorAccessibleGreen,
} from '@tds/core-colours'
import InputFeedback from '@tds/core-input-feedback'
import Paragraph from '@tds/core-paragraph'
import Text from '@tds/core-text'

import { borders } from '@tds/shared-styles'
import { safeRest } from '@tds/util-helpers'

import generateId from '../../shared/utils/generateId/generateId'
import ColoredTextProvider from '../../shared/components/ColoredTextProvider/ColoredTextProvider'

const FakeRadio = styled.span({
  height: '1.25rem',
  width: '1.25rem',
  minHeight: '1.25rem',
  minWidth: '1.25rem',

  outline: 0,
  lineHeight: 0,

  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  cursor: 'pointer',

  marginTop: '0.125rem',

  transition: 'border-color 0.1s linear, background-color 0.1s linear',
  borderRadius: '50%',
  ...borders.thin,

  borderColor: colorGreyShuttle,
  backgroundColor: colorWhite,
  '& > i': {
    display: 'none',
  },
})
const HiddenInput = styled.input({
  position: 'absolute',
  width: '1.2rem',
  height: '1.2rem',
  margin: '2px 1px',
  opacity: '0',
  pointerEvents: 'none',
})

const StyledLabel = styled.label(({ isError }) => ({
  display: 'flex',
  cursor: 'pointer',
  ...(isError && {
    color: colorCardinal,
    [`${HiddenInput}:checked ~ &`]: {
      color: 'initial',
    },
    [`div > ${FakeRadio}`]: {
      borderColor: colorCardinal,
    },
  }),
  [`${HiddenInput}:focus ~ & > div > ${FakeRadio}`]: {
    boxShadow: `0 0 4px 1px ${colorGreyShuttle}`,
    borderColor: isError ? colorCardinal : colorWhite,
  },
  [`${HiddenInput}:checked ~ & > div > ${FakeRadio}`]: {
    '& > span': {
      display: 'block',
    },
    borderColor: colorGreyShuttle,
  },
  [`${HiddenInput}:disabled ~ & > div > ${FakeRadio}`]: {
    borderColor: colorGreyGainsboro,
  },
  [`${HiddenInput}:disabled ~ & > div > div`]: {
    color: colorGreyGainsboro,
  },
}))

const InnerChecked = styled.span({
  width: '0.75rem',
  height: '0.75rem',
  borderRadius: '50%',
  backgroundColor: colorAccessibleGreen,
  display: 'none',
})

const StyledLabelAndDescriptionBox = styled(Box)({
  width: '100%',
})

const renderError = (error, errorId) => (
  <InputFeedback id={errorId} feedback="error">
    <Paragraph size="small">{error || ''}</Paragraph>
  </InputFeedback>
)

const getGeneratedId = (name, value) => {
  return generateId(name).postfix(value)
}
const getErrorId = (name, value, id) => {
  return generateId(id || getGeneratedId(name, value)).postfix('error-message')
}

/**
 * @version ./package.json
 */

const Radio = React.forwardRef(
  ({ id, name, value, label, feedback, error, description, ...rest }, ref) => (
    <Box between={2}>
      {feedback === 'error' && renderError(error, getErrorId(name, value, id))}
      <HiddenInput
        type="radio"
        id={id || getGeneratedId(name, value)}
        name={name}
        value={value}
        aria-invalid={feedback === 'error'}
        aria-describedby={feedback === 'error' ? getErrorId(name, value, id) : undefined}
        data-testid="hidden-input"
        ref={ref}
        {...safeRest(rest)}
      />
      <StyledLabel
        isError={feedback === 'error'}
        htmlFor={id || getGeneratedId(name, value)}
        data-testid="checkbox-label"
      >
        <Box between={3} inline>
          <FakeRadio data-testid="fake-input">
            <InnerChecked />
          </FakeRadio>
          <StyledLabelAndDescriptionBox between={2}>
            <ColoredTextProvider>
              <Text>{label}</Text>
            </ColoredTextProvider>
            {description && <Text size="small">{description}</Text>}
          </StyledLabelAndDescriptionBox>
        </Box>
      </StyledLabel>
    </Box>
  )
)

Radio.propTypes = {
  /**
   * The label.
   */
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  /**
   * Associate this radio with a group. Set as the HTML name attribute.
   */
  name: PropTypes.string.isRequired,
  /**
   * The value. Must be unique within the group.
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]).isRequired,
  /**
   * Use `checked` for controlled radio. For uncontrolled radio, use React's built-in `defaultChecked` prop.
   * See examples below for more details.
   */
  checked: PropTypes.bool,
  /**
   * The id. Must be unique within the group. If no id is provided, one will be generated in the following format: `name_value`
   */
  id: PropTypes.string,
  /**
   * Description text below the radio.
   */
  description: PropTypes.string,
  /**
   * A feedback state.
   */
  feedback: PropTypes.oneOf(['error']),
  /**
   * An error message.
   */
  error: PropTypes.string,
}

Radio.defaultProps = {
  id: undefined,
  description: undefined,
  feedback: undefined,
  error: undefined,
  checked: undefined,
}

Radio.displayName = 'Radio'

export default Radio
