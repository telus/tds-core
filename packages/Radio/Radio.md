### Minimal usage

Radio buttons are used when there is a list of two or more options that are mutually exclusive and the user must select exactly one choice. In other words, clicking a non-selected radio button will deselect whatever other button was previously selected in the list. Radio groups allow users to easily compare and see options.

<a href="https://www.nngroup.com/articles/checkboxes-vs-radio-buttons/" target="_blank">Reference</a>

### Usage criteria

- Use when there can only be exactly one choice from mulitple options
- Use there are 2-6 options to choose from
- Use [`Select`](#select) (dropdown) when there are over 6 options

```jsx
initialState = {
  choice: 'e.bill',
}

const setChoice = event => {
  setState({ choice: event.target.value })
}

<Box tag="fieldset" between={2}>
  <legend>
    <Text bold size="medium">
      How would you like to receive your monthly bill?
    </Text>
  </legend>
  <Radio
    label="e.Bill"
    name="monthly-bill"
    value="e.bill"
    checked={state.choice === 'e.bill'}
    onChange={setChoice}
  />
  <Radio
    label="Paper bill"
    name="monthly-bill"
    value="paper bill"
    checked={state.choice === 'paper bill'}
    onChange={setChoice}
  />
</Box>
```
