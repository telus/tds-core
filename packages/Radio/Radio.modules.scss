@import '../colour/colours';

.disabled {
  border-color: $color-gainsboro;
  background-color: $color-gainsboro;
}

.checked {
  border-color: $color-shuttle-grey;
  background-color: $color-white;

  cursor: pointer;
}

.disabledChecked {
  border-color: $color-gainsboro;
  background-color: $color-white;
}

.inner {
  composes: circular from '../../shared/styles/Borders.modules.scss';

  height: 12px;
  width: 12px;
}

.innerChecked {
  composes: inner;

  background-color: $color-accessible-green;
}

.innerDisabledChecked {
  composes: inner;

  background-color: $color-gainsboro;
}
