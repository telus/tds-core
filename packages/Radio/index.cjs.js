const Radio = require('./dist/index.cjs')

module.exports = Radio
