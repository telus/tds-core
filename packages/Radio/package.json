{
  "name": "@tds/core-radio",
  "version": "2.0.29",
  "main": "index.cjs.js",
  "module": "index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/telusdigital/tds.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "TELUS digital",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telusdigital/tds/issues"
  },
  "homepage": "http://tds.telus.com",
  "peerDependencies": {
    "react": "^16.8.2",
    "react-dom": "^16.8.2",
    "styled-components": "^4.1.3"
  },
  "dependencies": {
    "@tds/core-box": "^2.2.1",
    "@tds/core-colours": "^2.2.1",
    "@tds/core-input-feedback": "^2.0.18",
    "@tds/core-paragraph": "^2.0.12",
    "@tds/core-text": "^3.1.1",
    "@tds/shared-styles": "^1.5.2",
    "@tds/util-helpers": "^1.4.2",
    "prop-types": "^15.5.10"
  }
}
