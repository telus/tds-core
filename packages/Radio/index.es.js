import Radio from './dist/index.es'

export default Radio
