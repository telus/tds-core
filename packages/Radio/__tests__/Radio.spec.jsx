import React from 'react'
import { mount, render } from 'enzyme'

import Text from '@tds/core-text'
import DecorativeIcon from '@tds/core-decorative-icon'
import InputFeedback from '@tds/core-input-feedback'
import Radio from '../Radio'
import ColoredTextProvider from '../../../shared/components/ColoredTextProvider/ColoredTextProvider'

describe('Radio', () => {
  const defaultProps = {
    label: 'The radio',
    name: 'radio_group',
    value: 'the-value',
  }
  const doMount = (overrides = {}) => {
    const radio = mount(<Radio {...defaultProps} {...overrides} />)

    const findRadioElement = () => radio.find('input')

    return {
      radio,
      label: radio.find('label'),
      findRadioElement,
      findFakeRadio: () => radio.find('[data-testid="fake-input"]'),
      findFakeInnerRadio: () => radio.find('[data-testid="fake-inner-radio"]'),
      findColoredLabel: () => radio.find(ColoredTextProvider),
      findErrorMessage: () => radio.find(InputFeedback),
      check: () => findRadioElement().simulate('change', { target: { checked: true } }),
      focus: (focusEvent = {}) => findRadioElement().simulate('focus', focusEvent),
      blur: (blurEvent = {}) => findRadioElement().simulate('blur', blurEvent),
    }
  }

  it('renders', () => {
    const radio = render(<Radio label="A label" name="the-group" value="the-value" />)

    expect(radio).toMatchSnapshot()
  })

  it('allows numbers as value', () => {
    const radio = render(<Radio label="A label" name="the-group" value={1} />)

    expect(radio).toMatchSnapshot()
  })

  it('must have a label', () => {
    const { label } = doMount({ label: 'Some label' })

    expect(label).toContainReact(<Text size="medium">Some label</Text>)
  })

  it('must have a name and a value', () => {
    const { findRadioElement } = doMount({ name: 'some-radio-group', value: 'some-value' })

    expect(findRadioElement()).toHaveProp('name', 'some-radio-group')
    expect(findRadioElement()).toHaveProp('value', 'some-value')
  })

  it('has a fake radio', () => {
    const { findFakeRadio } = doMount()

    expect(findFakeRadio()).toHaveClassName('fakeRadio')
  })

  describe('connecting the label to the radio', () => {
    it('connects the label to the radio', () => {
      const { label, findRadioElement } = doMount()

      expect(label.prop('htmlFor')).toEqual(findRadioElement().prop('id'))
    })

    it('uses the id when provided', () => {
      const { label, findRadioElement } = doMount({
        id: 'the-id',
        name: 'the-radio-group',
        value: 'the-value',
      })

      expect(label).toHaveProp('htmlFor', 'the-id')
      expect(findRadioElement()).toHaveProp('id', 'the-id')
    })

    it('uses the name and the value when no id is provided', () => {
      const { label, findRadioElement } = doMount({ name: 'the-radio-group', value: 'the-value' })

      expect(label).toHaveProp('htmlFor', 'the-radio-group_the-value')
      expect(findRadioElement()).toHaveProp('id', 'the-radio-group_the-value')
    })
  })

  describe('interactivity', () => {
    it('can be unchecked', () => {
      const { findRadioElement, findFakeRadio, findFakeInnerRadio } = doMount({ checked: false })

      expect(findRadioElement()).toHaveProp('checked', false)
      expect(findFakeRadio()).toHaveClassName('unchecked')
      expect(findFakeInnerRadio()).not.toExist()
      expect(findFakeRadio().find(DecorativeIcon)).not.toExist()
    })

    it('can be checked', () => {
      const { findRadioElement, findFakeRadio, findFakeInnerRadio } = doMount({ checked: true })

      expect(findRadioElement()).toHaveProp('checked', true)
      expect(findFakeRadio()).toHaveClassName('checked')
      expect(findFakeInnerRadio()).toHaveClassName('innerChecked')
    })

    it('checks when clicking', () => {
      const { findRadioElement, findFakeRadio, findFakeInnerRadio, check } = doMount()

      check()

      expect(findRadioElement()).toHaveProp('checked', true)
      expect(findFakeRadio()).toHaveClassName('checked')
      expect(findFakeInnerRadio()).toHaveClassName('innerChecked')
    })

    it('notifies when it is checked', () => {
      const onChangeMock = jest.fn()
      const { check } = doMount({ onChange: onChangeMock })

      check()
      expect(onChangeMock).toHaveBeenCalledWith(
        expect.objectContaining({ target: { checked: true } })
      )
    })

    it('can receive a new value from a parent component', () => {
      const { radio, findRadioElement } = doMount({ checked: false })

      radio.setProps({ checked: true })

      expect(findRadioElement()).toHaveProp('checked', true)
    })
  })

  describe('focusing', () => {
    it('can be focused and unfocused', () => {
      const { findFakeRadio, focus, blur } = doMount()

      focus()
      expect(findFakeRadio()).toHaveClassName('focused unchecked')

      blur()
      expect(findFakeRadio()).not.toHaveClassName('focused')
      expect(findFakeRadio()).toHaveClassName('unchecked')
    })

    it('will notify when focus is gained', () => {
      const onFocusMock = jest.fn()
      const event = { target: { value: 'the value' } }

      const { focus } = doMount({ onFocus: onFocusMock })
      focus(event)

      expect(onFocusMock).toHaveBeenCalledWith(expect.objectContaining(event))
    })

    it('will notify when focus is lost', () => {
      const onBlurMock = jest.fn()
      const event = { target: { value: 'the value' } }

      const { blur } = doMount({ onBlur: onBlurMock })
      blur(event)

      expect(onBlurMock).toHaveBeenCalledWith(expect.objectContaining(event))
    })
  })

  describe('error', () => {
    it('can have an error feedback state', () => {
      const { findFakeRadio, findColoredLabel } = doMount({
        label: 'Some error',
        feedback: 'error',
      })

      expect(findColoredLabel()).toHaveProp('colorClassName', 'errorText')
      expect(findFakeRadio()).toHaveClassName('error')
      expect(findFakeRadio()).not.toHaveClassName('unchecked')
    })

    it('does not appear as an error when it is checked', () => {
      const { findFakeRadio, findColoredLabel, check } = doMount({
        label: 'Some error',
        feedback: 'error',
        checked: false,
      })

      check()

      expect(findColoredLabel()).not.toExist()
      expect(findFakeRadio()).toHaveClassName('checked')
      expect(findFakeRadio()).not.toHaveClassName('error')
    })
  })

  describe('disabled', () => {
    it('can be disabled', () => {
      const { findFakeRadio, findRadioElement, findColoredLabel } = doMount({
        label: 'A label',
        disabled: true,
      })

      expect(findColoredLabel()).toHaveProp('colorClassName', 'disabledText')
      expect(findRadioElement()).toHaveProp('disabled', true)
      expect(findFakeRadio()).toHaveClassName('disabled')
    })

    it('can be disabled and checked', () => {
      const { findFakeRadio, findFakeInnerRadio, findRadioElement, findColoredLabel } = doMount({
        label: 'A label',
        disabled: true,
        checked: true,
      })

      expect(findColoredLabel()).toHaveProp('colorClassName', 'disabledText')
      expect(findRadioElement()).toHaveProp('disabled', true)
      expect(findFakeRadio()).toHaveClassName('disabledChecked')
      expect(findFakeInnerRadio()).toHaveClassName('innerDisabledChecked')
    })
  })

  describe('accessibility', () => {
    it('marks the checkbox as invalid when in the error feedback state', () => {
      let findRadioElement = doMount().findRadioElement
      expect(findRadioElement()).toHaveProp('aria-invalid', false)

      findRadioElement = doMount({ feedback: 'error' }).findRadioElement
      expect(findRadioElement()).toHaveProp('aria-invalid', true)
    })

    it('does not attach aria-describedby to the checkbox when no error is present', () => {
      const { findRadioElement } = doMount({ error: undefined })

      expect(findRadioElement()).toHaveProp('aria-describedby', undefined)
    })

    it('connects the error message to the checkbox for screen readers', () => {
      const { findRadioElement } = doMount({
        id: 'some-field-id',
        error: 'An error message',
      })

      expect(findRadioElement()).toHaveProp('aria-describedby', 'some-field-id_error-message')
    })
  })

  it('passes additional attributes to the radio', () => {
    const { findRadioElement } = doMount({
      disabled: true,
      'data-some-attr': 'some value',
    })
    expect(findRadioElement()).toHaveProp('disabled', true)
    expect(findRadioElement()).toHaveProp('data-some-attr', 'some value')
  })

  it('does not allow custom CSS', () => {
    const { findRadioElement } = doMount({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(findRadioElement()).not.toHaveProp('className', 'my-custom-class')
    expect(findRadioElement()).not.toHaveProp('style')
  })
})
