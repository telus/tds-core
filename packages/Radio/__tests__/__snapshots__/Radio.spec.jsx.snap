// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio allows numbers as value 1`] = `
<div
  class="TDS_Box-modules__betweenBottomMargin-2___31zX_ TDS_Box-modules__stack___33m4D"
>
  <label
    for="the-group_1"
  >
    <span
      class="TDS_Box-modules__betweenRightMargin-3___1dOvx TDS_Box-modules__inline___jTHcz alignFlexStart"
    >
      <span
        class="fakeRadio unchecked"
        data-testid="fake-input"
      >
        <input
          aria-invalid="false"
          class="hiddenInput"
          id="the-group_1"
          name="the-group"
          type="radio"
          value="1"
        />
      </span>
      <div
        class="TDS_Box-modules__betweenBottomMargin-2___31zX_ TDS_Box-modules__stack___33m4D flexOne"
      >
        <span
          class="TDS_Typography-modules__medium___1rxfE TDS_Typography-modules__wordBreak___3OZx_ TDS_Typography-modules__mediumFont___XMrRj TDS_Typography-modules__color___2CNH8"
        >
          A label
        </span>
      </div>
    </span>
  </label>
</div>
`;

exports[`Radio renders 1`] = `
<div
  class="TDS_Box-modules__betweenBottomMargin-2___31zX_ TDS_Box-modules__stack___33m4D"
>
  <label
    for="the-group_the-value"
  >
    <span
      class="TDS_Box-modules__betweenRightMargin-3___1dOvx TDS_Box-modules__inline___jTHcz alignFlexStart"
    >
      <span
        class="fakeRadio unchecked"
        data-testid="fake-input"
      >
        <input
          aria-invalid="false"
          class="hiddenInput"
          id="the-group_the-value"
          name="the-group"
          type="radio"
          value="the-value"
        />
      </span>
      <div
        class="TDS_Box-modules__betweenBottomMargin-2___31zX_ TDS_Box-modules__stack___33m4D flexOne"
      >
        <span
          class="TDS_Typography-modules__medium___1rxfE TDS_Typography-modules__wordBreak___3OZx_ TDS_Typography-modules__mediumFont___XMrRj TDS_Typography-modules__color___2CNH8"
        >
          A label
        </span>
      </div>
    </span>
  </label>
</div>
`;
